lib/src/rules/invalid_case_patterns.dart

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2023, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+
+const _desc = r'Use case expressions that are valid in Dart 3.0.';
+
+/// todo(pq): add details or link out to a doc?
+const _details = r'''
+Some case expressions that are valid in Dart 2.19 and below will become an error
+or have changed semantics when a library is upgraded to 3.0. This lint flags
+those expressions in order to ease migration to Dart 3.0.
+''';
+
+class InvalidCasePatterns extends LintRule {
+  static const LintCode code = LintCode(
+      'invalid_case_patterns', 'Invalid case pattern.',
```

Comment:
I think it's reasonable to assume that users of this lint might not be familiar with the patterns feature, in which case I think this doesn't really help much. Minimally I think we want to say something like "This expression is not valid in a 'case' clause in Dart 3.0." Maybe with a correction message that suggests reading the lint's documentation for help migrating the code.